src/main/java/com/mongodb/spark/sql/connector/MongoCatalog.java

```diff
@@ -0,0 +1,422 @@
+   */
+  @Override
+  public void createNamespace(final String[] namespace, final Map<String, String> metadata)
+      throws NamespaceAlreadyExistsException {
+    assertInitialized();
+    if (namespaceExists(namespace)) {
+      throw new NamespaceAlreadyExistsException(namespace);
+    }
+  }
+
+  /**
+   * Apply a set of metadata changes to a namespace (database) in the catalog.
+   *
+   * <p>Throws {@code UnsupportedOperationException} as altering a namespace (database) is not
+   * supported.
+   *
+   * @param namespace (database) a multi-part namespace
+   * @param changes a collection of changes to apply to the namespace
+   * @throws UnsupportedOperationException If namespace (database) properties are not supported
+   */
+  @Override
+  public void alterNamespace(final String[] namespace, final NamespaceChange... changes) {
+    assertInitialized();
+    throw new UnsupportedOperationException("Altering databases is currently not supported");
+  }
+
+  /**
+   * Drop a namespace (database) from the catalog.
+   *
+   * @param namespace (database) a multi-part namespace
+   * @return true if the namespace (database) was dropped
+   */
+  @Override
+  public boolean dropNamespace(final String[] namespace) {
+    assertInitialized();
+    if (namespaceExists(namespace)) {
+      mongoConnectionProvider.doWithDatabase(namespace[0], MongoDatabase::drop);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * List the tables (collections) in a namespace (database) from the catalog.
+   *
+   * <p>If the catalog supports views, this must return identifiers for only tables (collections)
+   * and not views.
+   *
+   * @param namespace (database) a multi-part namespace
+   * @return an array of Identifiers for tables (collections)
+   */
+  @Override
+  public Identifier[] listTables(final String[] namespace) {
+    assertInitialized();
+    assert namespace.length == 1 : "Namespace size must equal 1";
+    return mongoConnectionProvider.withDatabase(
+        namespace[0],
+        db ->
+            db
+                .listCollections()
+                .filter(Filters.eq("type", "collection"))
+                .map(d -> Identifier.of(namespace, d.getString("name")))
+                .into(new ArrayList<>())
+                .stream()
+                .filter(identifier -> excludeSystemNamespaces(identifier.name()))
+                .toArray(Identifier[]::new));
+  }
+
+  /**
+   * Test whether a table (collection) exists using an {@link Identifier identifier} from the
+   * catalog.
+   *
+   * <p>If the catalog supports views and contains a view for the identifier and not a table, this
+   * must return false.
+   *
+   * @param identifier a table (collection) identifier
+   * @return true if the table (collection) exists, false otherwise
+   */
+  @Override
+  public boolean tableExists(final Identifier identifier) {
+    assertInitialized();
+    return Arrays.stream(listTables(identifier.namespace()))
```

Comment:
Could the server do the filtering for us?

Reply:
Done